replace int by enums in geometry add ring and add part

3nids opened the pull request:
Shall we use a single enum for all the methods or one enum per type of operation (as done in this PR) ?

Rephrasing is welcome ;)

3nids committed:
replace int by enums in geometry add ring and add part

3nids commented:
@mhugo would be nice to have your feedback since you wrote the TODOs

m-kuhn commented:
What do you think about having one shared enum for all functions?
I think it would be nice to be able to write the same everywhere

    if not addRing() == QgsGeometry.Success

3nids commented:
That's the question I'm asking in this PR ;)

There are pros and cons...
A single enum would make the code clearer, it avoids redundancy but it might get quite long.
Also one drawback is that it's a _hard_ API break since the old integer values won't correspond to the new enum anymore.

Personaly, I don't care. Would be a +0 for single enum.

m-kuhn commented:
We have been breaking worse than that ;)

3nids committed:
use single enum and add documentation

3nids commented:
@m-kuhn hardest break ever is in place. This seems indeed better this way.

wonder-sk commented:
If we use a single enum, it would be good to mention in doxygen documentation which are the possible result values. Otherwise the plugin developers either try to handle all errors even if not applicable - or they need to read code to see which ones to handle...

3nids committed:
fix test

3nids commented:
@wonder-sk that's a good point. Would you switch back to several enums?

wonder-sk commented:
Maybe not necessary to switch to several enums, rather to have one success value shared by all operations and then separate error values for each operation... something like:

enum { Success, AddRingErrorAbc, AddRingErrorDef, AddPartErrorAbc, ... }

3nids committed:
in OperationResult, separate members by type of operation, fix doxumentation

3nids commented:
OK will do.
Concerning splitPart, using an enum would mean changing up to QgsGeometryEngine (and QgsGeos).
Do you have any concern on this point?

wonder-sk commented:
Hmm for it I would probably use a separate enum within QgsGeometryEngine and remap them to QgsGeometry error values...

3nids committed:
splitPart also return QgsGeometry::OperationResult

QgsGeometryEngine::splitPart returns an enum: EngineOperationResult
a match is made between the two enums

m-kuhn commented on python/core/geometry/qgsgeometry.sip:
```diff
@@ -28,8 +28,25 @@ class QgsGeometry
+      /* General input issues */
+      InvalidInputType, /*!< The input geometry (ring or part or ...) has not the correct geometry type */
+      /* Ring specific issues*/
+      RingNotClosed, /*!< The imput ring is not closed */
```
imput -> input

m-kuhn commented:
> Maybe not necessary to switch to several enums, rather to have one success value shared by all operations and then separate error values for each operation... something like:
> 
> enum { Success, AddRingErrorAbc, AddRingErrorDef, AddPartErrorAbc, ... }

Not sure about the advantage of this... I'd rather go for either share values wherever possible (and skip the `AddPart`-part) and document methods properly or go for multiple enums.

3nids committed:
reshape returns QgsGeometry::OperationResult

3nids commented:
I added the changes to QgsGeos and QgsGeometryEngine.
I have no clear preference for one or several enums, waiting on other's opinion

3nids committed:
fix test

3nids committed:
fix documentation

nyalldawson commented:
My vote would go to one enum... Having multiple just seems like a recipe for inconsistent naming

m-kuhn commented:
Superseded by https://github.com/qgis/QGIS/pull/5014
